### ISSUE TITLE
Redefinition of CellAccessor::active_cell_index()

### ISSUE THREAD

**Issue by peterrum** [dealii/dealii#10530]
PR #10490 introduces globally cell unique indices for active and multigrid levels for locally owned and ghost cells. Furthermore, it introduces partitioners. The global indices and the partitioners could be used to redefine consistently `active_cell_index()`. A code might look like this:
```cpp
template <int dim, int spacedim>
unsigned int
CellAccessor<dim, spacedim>::active_cell_index() const
{
  const auto i = this->tria->levels[this->present_level]
                   ->global_cell_indices[this->present_index];

  const auto tria_parallel =
    dynamic_cast<const parallel::TriangulationBase<dim, spacedim> *>(
      &*this->tria);

  if (tria_parallel && (tria_parallel->global_cell_index_partitioner().size() > 0))
    return (tria_parallel->global_cell_index_partitioner().global_to_local(i));
  else
    return i;
}
````
the values could be of course be pre-computed. This way to define `active_cell_index()`  would enable to write into local vector (only consisting of locally owned cells - as needed in the case of error estimation) and global vectors (with ghost layers).

An incompatible change would be that artificial cells would not get an `active_cell_index`. In my opinion this would be not a problem, in contrast: in most cases, it gives you an indication that you are doing something wrong.

Furthermore, the following code would not work anymore:
```cpp
template <int dim>
void
check(const Triangulation<dim> &tria)
{
  unsigned int index = 0;
  for (typename Triangulation<dim>::active_cell_iterator cell =
         tria.begin_active(); cell != tria.end(); ++cell, ++index)
    Assert(cell->active_cell_index() == index, ExcInternalError());
}
```
which should not be a big problem since `active_cell_index()` is currently mostly used to access local vectors. As long as `active_cell_index()` is consistently used to write into the vector and to read from it, it should be fine. See also the discussion in #10528.

What do others think? I will prepare I bunch of PRs with minor adjustments to make the change proposed in this PR possible and will use this issue page to mange those PRs.

**Comment by bangerth** [comment]
There are quite a lot of places right now where we create vectors of size `n_active_cells()` (which includes ghost and artificial cells) and then only consider those parts of the vector that correspond to locally owned cells. Error estimation, refinement, data out, etc, all fall into this category.

Your change would make incompatible changes to these places, and I don't actually think this is necessary. In particular, the current workflow keeps everything local: None of these functions ever need to exchange information with other processors, and so it seems entirely appropriate to let every processor generate its own vector, rather than have one global vector into which we index. Making such a change would introduce synchronization points for no actual benefit.

In other words, I think that the current design is actually the better one. I don't see a reason why we ought to change it.

**Comment by tjhei** [comment]
> In other words, I think that the current design is actually the better one. I don't see a reason why we ought to change it.

I agree that there are good reasons to keep things as is (including not doing unnecessary user-facing changes), but I also think the current design is not ideal:
1. Exposing artificial cells is not particularly useful. We should have never included them in n_active_cells, active_cell_index, iterators, etc..
2. I find the current design using local vectors for error estimation quite confusing, especially when computing face terms (see [1] for example)

So, I don't know if this is something worth looking into or not. Making changes is hard and I regret many design choices we did years ago...

[1]: https://github.com/dealii/dealii/blob/301fce157e88add9a6a41e307d7cc5767b679cf1/examples/step-50/step-50.cc#L1382-L1387

**Comment by bangerth** [comment]
This particular mistake is not high on my list of things I lose sleep over ;-) It's worth pointing out that `cell->active_cell_index()` came only after this design choice for local vectors. Before we just visited all cells in order and needed a vector of the same number of cells we were visiting, where we just counted an index up by one for each cell.

It is whatever it is -- I think the current design is neither particularly inefficient nor hard to understand. I would suggest to leave things as they are. If you need indices for other functionality, there is no reason not to add more pieces of information to the triangulation.

**Comment by peterrum** [comment]
@bangerth Let me shortly explain the background of my suggestion and the current status.

Before that, a comment related to
>  I regret many design choices we did years ago...

> This particular mistake is not high on my list of things I lose sleep over ;-)

This issue is not about mistakes made in the past. It is more about future maintainability.



With #10490, @kronbichler and I would like to introduce a global numbering of active and level cells. One of the reasons behind this PR is to be able (to query a unique index in a global contiguous index set and) to create global cell-data vector without the need to create `DoFHandler` with `FE_DGQ(0)`.



In the first implementation, each cell actually stored a new global index. However, @tjhei pointed out in https://github.com/dealii/dealii/pull/10490#issuecomment-643485064 the similarity to `active_cell_index()`. He also suggested to get rid of this function, however I used the comments to think about a way to keep `active_cell_index()` and to think about what needs to be changed in that function so that it can be used together with a `Partitioner` to create a global version `active_cell_index()` aka. `global_active_cell_index()`. PR #10490 shows an implementation of `global_active_cell_index()` based on `active_cell_index()`:
```cpp
template <int dim, int spacedim>
inline types::global_cell_index
CellAccessor<dim, spacedim>::global_active_cell_index() const
{
  // get local index
  const types::global_cell_index local_index = this->active_cell_index();

  if (local_index == numbers::invalid_dof_index)
    return numbers::invalid_dof_index;

  // for distributed triangulations: translate local to global index
  if (const auto tria_parallel = dynamic_cast<
        const parallel::DistributedTriangulationBase<dim, spacedim> *>(
        &*this->tria))
    return tria_parallel->global_cell_index_partitioner().local_to_global(
      local_index);

  // for serial and shared triangulation: simply return local index
  return local_index;
}

```
(the code in the issue description is outdated).

In the context of local vectors the usage would not change:
```cpp
vec[cell->active_cell_index()];
```
in the context of global vectors, one would have two options:
```
vec.local_element (cell->active_cell_index());
vec[cell->active_cell_index()];
```

The changes I had to make to make this code work is:

- to not assign active indices to artificial cells

- first enumerate locally owned cells and after that ghost cells.


This change of course led to some failing tests. However, these could be fixed by consistent use of `active_cell_index()`. You pointed out the error estimator. However, it turned out if the error vector is filled and read consistently everything is fine (in particular artificial cells are never processed!).


I am quite happy with the current design (that the local (modified) active cell is used to create a global active cell) and I am glad that @tjhei pointed out the similarity. The consequence of the design is that we actually don't get a new index/id, but it is directly liked to a currently existing one. We already have many versions of indices/ids (see: https://github.com/dealii/dealii/pull/10490#issuecomment-643381430). A completely/separate new type of index would only lead to confusion for users (which is why I wanted to make the global coarse id private shortly before release).

Let's turn back to my comment related to `future maintainability`. I would much more prefer to only have one type of active index (+ partitioner), instead of maintaining a second one. Having said that, it is clear to me that there would for now two types of active indices for compatibility reasons until we have deprecated/removed functions which take the old index order.
